json-transport/src/lib.rs

```diff
@@ -149,36 +130,40 @@ where
     Item: for<'de> Deserialize<'de>,
     SinkItem: Serialize,
 {
-    Ok(new(TcpStream::connect(addr).compat().await?))
+    Ok(new(TcpStream::connect(addr).await?))
 }
 
 /// Listens on `addr`, wrapping accepted connections in JSON transports.
-pub fn listen<Item, SinkItem>(addr: &SocketAddr) -> io::Result<Incoming<Item, SinkItem>>
+pub async fn listen<A, Item, SinkItem>(addr: A) -> io::Result<Incoming<Item, SinkItem>>
 where
+    A: ToSocketAddrs,
     Item: for<'de> Deserialize<'de>,
     SinkItem: Serialize,
 {
-    let listener = TcpListener::bind(addr)?;
+    let listener = TcpListener::bind(addr).await?;
     let local_addr = listener.local_addr()?;
-    let incoming = listener.incoming().compat();
+    let incoming = Box::pin(listener.incoming());
     Ok(Incoming {
         incoming,
         local_addr,
         ghost: PhantomData,
     })
 }
 
+trait IncomingTrait: Stream<Item = io::Result<TcpStream>> + std::fmt::Debug + Send {}
```

Comment:
Nit: this is only used in one place currently. Will it be used elsewhere in the future? If not, consider inlining.

Reply:
Required for the trait object below.